### PR TITLE
Connection limits per IP for WebSocket server

### PR DESCRIPTION
**Pull request by Ivo-Yankov**
**Description**:
- Adds a configurable limit on the maximum connections from a single IP address.
- Adds a class for WebSocket errors. I suggest we follow the convention described in this [article](https://medium.com/weekly-webtips/the-essential-guide-on-the-web-socket-status-codes-97b54773f66d) regarding the error codes. E.g. our error codes should look like `4XXX`

**Related issue(s)**:

Addresses #995 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: .env.example
Comment (Nana-EC) [inline .env.example:39]: nit: let's differentiate the servers
```suggestion
WS_CONNECTION_LIMIT_PER_IP=
```

File: packages/server/tests/acceptance/ws/subscribe.spec.ts
Comment (Nana-EC) [inline packages/server/tests/acceptance/ws/subscribe.spec.ts:294]: I'm weary of waits in test code but not sure how else to achieve this now

File: packages/server/tests/acceptance/ws/subscribe.spec.ts
Comment (Nana-EC) [inline packages/server/tests/acceptance/ws/subscribe.spec.ts:304]: `closeEventHandled` is never reset to false.
How would a failure when `i > 0` occur?

File: packages/ws-server/src/webSocketServer.ts
Comment (Nana-EC) [inline packages/ws-server/src/webSocketServer.ts:73]: This should be from a constant so you're not looking up and converting each time.
If you're not exposing publicly not sure that a getter is needed

File: packages/ws-server/src/webSocketServer.ts
Comment (Nana-EC) [inline packages/ws-server/src/webSocketServer.ts:82]: Q: should the decrement be moved to after the actual unsubscribe is done?
What happens if unsubscribe has a failure, does it still unsubscribe but report the failure or could there be a case where the subscription or connection exists but we decremented?

File: packages/ws-server/src/webSocketServer.ts
Comment (Nana-EC) [inline packages/ws-server/src/webSocketServer.ts:114]: we should have metrics when this occurs

File: packages/ws-server/src/webSocketServer.ts
Comment (Nana-EC) [inline packages/ws-server/src/webSocketServer.ts:77]: Q: Should we use a similar approach tot he relay where we pulled out the rate limiting logic into a separate class instead of lumping it in the main server class?
I think this would be cleaner and easier to maintain.

**Review by Nana-EC — COMMENTED** [review]
Looking good, some questions

File: packages/ws-server/src/webSocketServer.ts
Comment (ebadiere) [inline packages/ws-server/src/webSocketServer.ts:62]: I agree that this approach is cleaner for handling the connection limits.  I would like to wait until this is merged into main so I can pick it up in this [PR](https://github.com/hashgraph/hedera-json-rpc-relay/pull/1017) and use it.

File: packages/ws-server/src/webSocketServer.ts
Comment (Ivo-Yankov) [reply packages/ws-server/src/webSocketServer.ts:82]: Good point, `unsubscribe` should be executed before all decrements.

File: packages/ws-server/src/webSocketServer.ts
Comment (Ivo-Yankov) [reply packages/ws-server/src/webSocketServer.ts:114]: There is an issue for adding metrics to the ws-server, with a bullet point of tracking the count of active connections by IP

**Review by Nana-EC — APPROVED** [review]

File: packages/ws-server/src/ConnectionLimiter.ts
Comment (Nana-EC) [inline packages/ws-server/src/ConnectionLimiter.ts:70]: nit: still think this should be a constant at the class level. Maybe even on instantiation of this file

Comment (Ivo-Yankov) [reply packages/ws-server/src/ConnectionLimiter.ts:70]: I did it this way so that it's possible to change the value after initialization - this is used in the tests. I know it's not the best solution, but the alternative is to add setter/getter methods and I think that would add unnecessary complexity.

**Pull request merged**
